### ISSUE TITLE
kundera.client missing with external configuration

### ISSUE THREAD

**Issue by msfbrasil** [Impetus/kundera#526]
Hello there!

I'm using Kundera 2.9 on an application to access a Cassandra database.
Everything was completely fine before change below:

I'm migrating all properties present on "persistence.xml" to an external file and getting them set on "EntityManagerFactory" instantiation.

Code is below:
# 

Map<String, String> propertyMap = new HashMap<String, String>();

propertyMap.put(CassandraConstants.CQL_VERSION, CassandraConstants.CQL_VERSION_3_0);

propertyMap.put(KUNDERA_NODES_PERSISTENCE_PROPERTY, kunderaNodes);
propertyMap.put(KUNDERA_PORT_PERSISTENCE_PROPERTY, kunderaPort);
propertyMap.put(KUNDERA_KEYSPACE_PERSISTENCE_PROPERTY, kunderaKeyspace);
propertyMap.put(KUNDERA_DIALECT_CONFIG, kunderaDialect);
propertyMap.put(KUNDERA_CLIENT_LOOKUP_CLASS_CONFIG, kunderaClientLookupClass);

this.entityManagerFactory = Persistence.createEntityManagerFactory(PERSISTENCE_UNIT, propertyMap);
# 

The problem is that I'm getting the error: "com.impetus.kundera.metadata.model.PersistenceUnitMetadata.getClient - kundera.client property is missing for persistence unit:macs-db".

Getting into Kundera codes, I've noticed that "MetamodelConfiguration" calls a "getClient" method below:

String clientFactoryName = KunderaMetadataManager.getPersistenceUnitMetadata(m.getPersistenceUnit())
                .getClient();

And on this method, the map used to check the existence of "kundera.client.lookup.class" property has no information.
These properties are visible at various points of the code, but not at this place.

Any ideas?
Is Kundera able to accept external configuration at this version?

Thanks and best regards,
Mauro.

**Comment by msfbrasil** [comment]
Hello there!

Another peace of information that could be useful is the stack prior to exception thrown regarding the missing configuration, that is:

# 

PersistenceUnitMetadata.getClient() line: 540
MetamodelConfiguration.processGeneratedValueAnnotation(Class<?>, String, EntityMetadata, Map<String,IdDiscriptor>) line: 506
MetamodelConfiguration.scanClassAndPutMetadata(InputStream, Reader, Map<String,EntityMetadata>, Map<String,Class<?>>, String, String, Map<String,List<String>>, Map<String,IdDiscriptor>) line: 406
MetamodelConfiguration.loadEntityMetadata(String) line: 232
MetamodelConfiguration.configure() line: 103
Configurator.configure() line: 65
KunderaPersistence.initializeKundera(String, Map) line: 109 

# 

I'm not completely sure if this is important to mention. If this could lead to a flux of events that result on this exception, but the fact is that I'm not defining the entities classes at the "persistence.xml" file.

Thanks and best regards,
Mauro.

**Comment by msfbrasil** [comment]
Hello there!

One place that externally provided properties could be "merged" to "loaded by file" properties of "PersistenceUnitMetadata" definitions coud be at "loadEntityMetadata" method right after code block below:

# 

PersistenceUnitMetadata puMetadata = persistentUnitMetadataMap.get(persistenceUnit);
classesToScan = puMetadata.getManagedClassNames();
managedURLs = puMetadata.getManagedURLs();
Map<String, Object> externalProperties = KunderaCoreUtils.getExternalProperties(persistenceUnit, externalPropertyMap, persistenceUnits);

# 

"Merging" the "externalPropertyMap" properties information with "PersistenceUnitMetadata"s information (given by "puMetadata" above) at this point would avoid the exception got (I think, given that I didn't effectively tested the change).

Just a suggestion!
Hope this helps.

Thanks and best regards,
Mauro.

**Comment by mevivs** [comment]
Hi,
I will have a look and get back to you in few minutes.

-Vivek

**Comment by mevivs** [comment]
Given below code snippet looks problematic to me:

```
    private void processGeneratedValueAnnotation(Class<?> clazz, String persistenceUnit, EntityMetadata m,
            Map<String, IdDiscriptor> entityNameToKeyDiscriptorMap)
    {
        GeneratedValueProcessor processer = new GeneratedValueProcessor();
        String pu = m.getPersistenceUnit()/* getPersistenceUnitOfEntity(clazz) */;
        String clientFactoryName = KunderaMetadataManager.getPersistenceUnitMetadata(kunderaMetadata,
                m.getPersistenceUnit()).getClient();
        if (pu != null && pu.equals(persistenceUnit)
                || clientFactoryName.equalsIgnoreCase("com.impetus.client.rdbms.RDBMSClientFactory"))
        {
            Field f = (Field) m.getIdAttribute().getJavaMember();

            if (f.isAnnotationPresent(GeneratedValue.class))
            {
                processer.process(clazz, f, m, entityNameToKeyDiscriptorMap);
            }
        }
    }

```

In case of auto id generation, this would happen. However it should work fine  with non @GeneratedValue entities.

Fix for this issue is:

```
    private void processGeneratedValueAnnotation(Class<?> clazz, String persistenceUnit, EntityMetadata m,
            Map<String, IdDiscriptor> entityNameToKeyDiscriptorMap)
    {
        GeneratedValueProcessor processer = new GeneratedValueProcessor();
        String pu = m.getPersistenceUnit()/* getPersistenceUnitOfEntity(clazz) */;

        Map<String, Object> externalProperties = KunderaCoreUtils.getExternalProperties(persistenceUnit,
                externalPropertyMap, persistenceUnits);

        String clientFactoryName  = externalProperties != null ? (String) externalProperties
                .get(PersistenceProperties.KUNDERA_CLIENT_FACTORY) : null;

        if (clientFactoryName == null)
        {
            clientFactoryName = KunderaMetadataManager.getPersistenceUnitMetadata(kunderaMetadata,
                    m.getPersistenceUnit()).getClient();
        }

        if (pu != null && pu.equals(persistenceUnit)
                || clientFactoryName.equalsIgnoreCase("com.impetus.client.rdbms.RDBMSClientFactory"))
        {
            Field f = (Field) m.getIdAttribute().getJavaMember();

            if (f.isAnnotationPresent(GeneratedValue.class))
            {
                processer.process(clazz, f, m, entityNameToKeyDiscriptorMap);
            }
        }
    }
```

I will check-in this fix after some refactoring. But i guess it should work for you

-Vivek

**Comment by msfbrasil** [comment]
Hello Vivek, and thanks for the help!

The problem is at "getClient" method called on line below:

String clientFactoryName = KunderaMetadataManager.getPersistenceUnitMetadata(kunderaMetadata,
                m.getPersistenceUnit()).getClient();

Inside method "processGeneratedValueAnnotation".

The "getClient" validates the existence of "kundera.client.lookup.class" property on "PersistenceUnit" instance, that don't have the properties that weren't load from "persistence.xml".

Thanks and best regards,
Mauro.

**Comment by msfbrasil** [comment]
Just one little peace of information.

I'm considering here that something will need to be changed (and a new version released), because the dynamic configuration is "not" working here because of the problem reported above.

Please, let me know If I'm doing something wrong.

Thanks a lot and best regards,
Mauro.

**Comment by mevivs** [comment]
As mentioned above, it needs a fix and i will post an update once check-in this in current trunk

**Comment by msfbrasil** [comment]
I was just clarifying that it don't work for me (because of your comment "But i guess it should work for you"). Just that.

I'll wait the fix.

Thanks a lot and best regards.

**Comment by mevivs** [comment]
A fix has been added for this on current trunk. Please verify with latest snapshot.

-Vivek

**Comment by msfbrasil** [comment]
I'm getting this error while building "2.11-SNAPSHOT":

[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:2.5.1:testCompile (default-testCompile) on project kundera-cassandra: Compilation failure
[ERROR] /home/ima_mbrasil/projects/github/Kundera/Kundera/src/kundera-cassandra/src/test/java/com/impetus/client/crud/BaseTest.java:[50,20] error: cannot find symbol
[ERROR] -> [Help 1]

Any ideas?

Thanks and best regards,
Mauro.

**Comment by mevivs** [comment]
You need to manually build snapshot

-Vivek

**Comment by msfbrasil** [comment]
Tested and working pretty fine.

Thanks and best regards,
Mauro.

**Comment by vanessaschissato** [comment]
He guys,

I'm using Kundera 2.9 and setting the properties externally to pesistence.xml, like msfbrasil:

Map propertyMap = new HashMap();
...
propertyMap.put("kundera.client.config", "KunderaLoadConfig.xml"); 

this.entityManagerFactory = Persistence.createEntityManagerFactory(PERSISTENCE_UNIT, propertyMap);

BUT, doesn't work the "try next on fail" definied in KunderaLoadConfig.xml unless all the properties are configured in persistence.xml directly.

**Comment by mevivs** [comment]
```
BUT, doesn't work the "try next on fail" definied in KunderaLoadConfig.xml unless all the properties are configured in persistence.xml directly.
```

I did not get you on this. Any example?

-Vivek

**Comment by msfbrasil** [comment]
Hello there!

Maybe I can be of some help on that, given that this is an evolution of our internal development.

At some point, we have made the configuration loading of most of properties from an external properties file.
Providing these properties via a map on "EnittyManagerFactory" initialization, like show below:

---

Map<String, String> propertyMap = new HashMap<String, String>();

propertyMap.put(CassandraConstants.CQL_VERSION, CassandraConstants.CQL_VERSION_3_0);

propertyMap.put(KUNDERA_NODES_PERSISTENCE_PROPERTY, kunderaNodes);
propertyMap.put(KUNDERA_PORT_PERSISTENCE_PROPERTY, kunderaPort);
propertyMap.put(KUNDERA_KEYSPACE_PERSISTENCE_PROPERTY, kunderaKeyspace);
propertyMap.put(KUNDERA_DIALECT_CONFIG, kunderaDialect);
propertyMap.put(KUNDERA_CLIENT_LOOKUP_CLASS_CONFIG, kunderaClientLookupClass);

this.entityManagerFactory = Persistence.createEntityManagerFactory(PERSISTENCE_UNIT, propertyMap);

---

After some time, a colleague of mine started testing the application on production with a real Cassandra cluster.
First fail-over tests showed problems given that Kundera seemed to not notice that a Cassandra node was made unavailable.

After some research she found out that some extra configuration must be provided on a different ".xml" file and configured on "persistence.xml" through property "kundera.client.config".

The problem she founds was:

1- If all configuration is on "persistence.xml", including those that were already been loaded from an external property file, everything works fine. The client notices that Cassandra node is out and perform the fail-over perfectly.

2- But, if we keep loading the general configurations (nodes, dialect, client, etc) from an external property file and building the EntityManagerFactory configuration map dynamically, it doesn't matter if we put "kundera.client.config" property on "persistence.xml" or load it dynamically, the fail over doesn't work anymore.

Maybe the additional client configuration file is not getting loaded when the main configuration - nodes, dialect, and mainly the client iself - is provided through the map?

Thanks and best regards,
Mauro.

**Comment by mevivs** [comment]
```
2- But, if we keep loading the general configurations (nodes, dialect, client, etc) from an external property file and building the EntityManagerFactory configuration map dynamically, it doesn't matter if we put "kundera.client.config" property on "persistence.xml" or load it dynamically, the fail over doesn't work anymore.
```

Is it still an issue with latest trunk? I think it should load external config file property dynamically.

External configuration file are for "data store specific" configurations. For more details on this, please refer https://github.com/impetus-opensource/Kundera/wiki/Cassandra-Specific-Features

@msfbrasil 
Using Kundera in production? If yes, can we publish it over https://github.com/impetus-opensource/Kundera/wiki/Kundera-in-Production-Deployments? Would be great if you could drop an email regarding your usage of Kundera at kundera@impetus.co.in

Cheers,
-Vivek

**Comment by vanessaschissato** [comment]
The tests were done with versions 2.9 e 2.10.
I will try with the trunk code.

**Comment by mevivs** [comment]
Fixed with 2.11
